Add Swipe to Carousel Elements

AlexMNet opened the pull request:
Add the ability to swipe left/right on carousel elements on mobile sized devices using the Hammer JS Library
Hammer JS Library installed as well as the HammerJS Types for typescript.

Commit:
Add Swipe to carousel on mobile

Add the ability to swipe on carousel elements on mobile devices using
the Hammer JS Library.

Commit:
Merge branch 'main' into alexmnet-swipe-carousels
Update branch with main branch most current changes

Commit:
remove comma typo in carousel ts

AlexMNet commented:
It's easy to accidentally scroll up or down the page when trying to swipe through the carousel. Need to look into fixing that.

mckoss commented on src/carousel.ts:
```diff
@@ -22,6 +24,34 @@ function carousel(id: string, curChild=0) {
 
+
+    //Enable Swipe on mobile devices only
+    hammer.set({enable: windowSize <= 768 ? true : false})
```
Don't we want this to work on an iPad too?

mckoss committed:
Moved import - improve lint checks

mckoss commented on src/carousel.ts:
```diff
@@ -1,10 +1,14 @@
+import Hammer from 'hammerjs';
```
This is the correct place to import.  We generally don't want to depend on global variables for imports.

mckoss commented on src/carousel.ts:
```diff
@@ -22,6 +26,34 @@ function carousel(id: string, curChild=0) {
+        if(isFinal){
+            //Check if user swipes right (image moves right)
+            if(type === 'swiperight'){
+                if(curChild === 0){
```
We are getting a lot of redundant code here.  Would be best to consolidate.  Perhaps replace the addEventListeners on divLeft and divRight with:

() => scrollBy(-1)  and () => scrollBy(1)

```
function scrollBy(n: number) {
  const nextChild = curChild + n;
  if (nextChild < 0 || nextChild > numChildren) {
     return;
  }
  curChild = nextChild;
  centerCard(curChild);
 showControls();
```
  
}

mckoss commented on src/carousel.ts:
```diff
@@ -22,6 +26,34 @@ function carousel(id: string, curChild=0) {
+    //Enable Swipe on mobile devices only
+    hammer.set({enable: windowSize <= 768 ? true : false})
+    //Detect swipe left or right on divParent
+    hammer.on("swipeleft swiperight", function(e: any) {
```
Prefer arrow function:

(e) => {

any is not a great type declaration.  Hammer should be typed in such a way as to
not need an explicit declaration here - it should be inferred and have Hammer's
properties added to the event?

mckoss commented on src/carousel.ts:
```diff
@@ -22,6 +26,34 @@ function carousel(id: string, curChild=0) {
+            }
+            //Check if user swipes left (image moves left)
+            if(type === 'swipeleft'){
+                if(curChild === numChildren - 1){
```
scrollBy(1)

mckoss committed:
Fix missing semicolons.

mckoss reviewed (changes requested):
I've added some lint checks too for missing semicolons.  And fixed the files that had errors.

mckoss committed:
Enforce spacing around keywords - like if

mckoss commented on src/carousel.ts:
```diff
@@ -22,6 +26,34 @@ function carousel(id: string, curChild=0) {
+        const {type, isFinal} = e;
+
+        //Event listener fires mutiple times. Waits until events are done.
+        if(isFinal){
```
I want to enforce space after "if" - I added a lint rule to pick these up.

mckoss commented:
@AlexMNet I've left some review comments - and added some lint rules that are not currently passing.

You'll have to pull this branch and then continue to pick up my changes to your branch.

Commit:
Review Changes

Add scrollBy function to reduce redunant code.
Change callback in hammer function to arrow.
Increase pixel size to 1024 to allow swipe on mobile and tablets

Commit:
Merge branch 'main' into alexmnet-swipe-carousels
Update branch with latest in main

AlexMNet commented:
@mckoss Changes have been made!

mckoss commented on src/carousel.ts:
```diff
@@ -22,21 +26,25 @@ function carousel(id: string, curChild=0) {
-        if (curChild === numChildren - 1) {
-            return;
+    //Enable Swipe on mobile/tablet devices only
+    hammer.set({enable: windowSize <= 1024 ? true : false});
```
Why not get rid of this altogether?  And aren't some iPad dimensions over 2000 pixels?  Not to mention other devices with touch screens.  This doesn't hurt to have this library loaded on all devices, does it?

AlexMNet replied:
It actually works quite well on desktop by clicking and dragging. Only issue is if a user tries to click and drag on the image it drags a ghost image and it doesn't swipe. I believe the ghost image issue should be an easy fix though and can create another issue for that and handle.

mckoss committed:
Don't allow naked if bodies - fix import location of types/hammerjs

mckoss reviewed (changes requested):
I made a few edits again.  We don't allow if's w/o curly body as a style convention (I added another eslint rule to catch it).

You'll have to pull again.

Also - types are dev dependencies, not runtime dependencies.  I fixed the TS error.

Last issue is removing the screen width check if you think that's ok.

Commit:
Allow swipe on all devices

Remove check for screen size to allow swipe on all devices.

mckoss commented on src/carousel.ts:
```diff
@@ -22,22 +25,24 @@ function carousel(id: string, curChild=0) {
-    divRight.addEventListener('click', () => {
-        if (curChild === numChildren - 1) {
-            return;
+        //Event listener fires multiple times. Waits until events are done.
```
Space after comment //

mckoss commented on src/carousel.ts:
```diff
@@ -53,4 +58,16 @@ function carousel(id: string, curChild=0) {
+
+    function scrollBy(n: number) {
+        const nextChild = curChild + n;
+        console.log("nuChildren:", numChildren);
```
Don't leave console.log's in

mckoss committed:
Fix space after comment

mckoss reviewed (approved):
I'll merge this one - good job!
